DIPs/DIP-1NN4-QFS.md

```diff
@@ -0,0 +1,573 @@
+* a possibly qualified static array of essentially `T` — or
+* a possibly qualified associative array with value type an essentially `T`.
+
+In this DIP, *FP/D (type)* will be an abbreviation for *function pointer (type) or delegate (type)*.
+Also, *eFP/D (type)* will be used for *essentially an FP/D (type)*.
+
+An *essential call* of an eFP/D means an expression that is: In case of the eFP/D being
+* a possibly qualified FP/D: A call (in the regular sense) to that object.
+* a pointer `p` to an eFP/D: An essential call to `*p`.
+* a slice or a static or associative array `arr` to an eFP/D: An essential call to `arr[i]` for a suitable index `i`.
+
+This document makes use of the terms *(function) parameter* and *(function) argument* in a very precise manner.
+As a reminder for the reader, a parameter is a variable declared at a specific place,
+namely the function signature;
+an argument is an expression on the uppermost level in a function call.
+
+A *higher-order function*, or *functional* for short, is anything that can be called
+(including any kind of functions, function pointers, delegates, and `opCall`)
+that takes one or more eFP/D types as arguments.
+
+When a higher-order function is called, there are three notable entities to commonly refer to:
+* The *context function,* or *context* for short, is the function that contains the call expresion.
+* The *functional* is the higher-order function that is called.
+* The *parameter functions* are the variables declared by the functional's parameter list.
+* The *argument functions* or *callbacks* are the values plugged-in in the call expression.
+
+Although not an entity in the above sense, the functional's *parameter types* will be commonly referred to as such.
+
+An illustration of the last terms given in this code snippet using telling identifiers:
+```D
+alias ParameterType = void function();
+void functional(ParameterType parameterFunction)
+{
+    parameterFunction();
+}
+
+void context()
+{
+    alias callback = { };
+    functional(callback);
+}
+```
+
+## Rationale
+<!--
+A short motivation about the importance and benefits of the proposed change.
+An existing, well-known issue or a use case for an existing projects can greatly increase the
+chances of the DIP being understood and carefully evaluated.
+-->
+
+Higher-order functions are a disruption in a warrant attribute context.
+The execution of a context function may provably satisfy the conditions of a warrant attribute,
+but is considered illegal because the formal conditions of the warrant attribute not met.
+
+Warrant attributes on a functional (or, in fact, on any function) and on a functional's parameter mean very differnt things:
+* On a function, they give rise to a *guarantee* the function makes. E.g. a `@nogc` function will not allocate GC memory.
+(When a function returns a delegate or function pointer, warrant attributes on its return type are a guarantee of that function, too.)
```

Comment:
I don't understand this sentence. You can have a `@nogc` function returning a non-`@nogc` delegate.

Reply:
If the last one was intended to be a question, the answer is yes.
I should delete the "E.g." sentence and the one in parentheses. No real value here.